File: document/src/vespa/document/update/assignfieldpathupdate.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "assignfieldpathupdate.h"
#include <vespa/document/fieldvalue/fieldvalues.h>
#include <vespa/document/fieldvalue/iteratorhandler.h>
#include <vespa/document/select/parser.h>
#include <vespa/document/select/variablemap.h>
#include <vespa/document/serialization/vespadocumentdeserializer.h>
#include <vespa/vespalib/objects/nbostream.h>
#include <vespa/vespalib/util/exceptions.h>
#include <boost/numeric/conversion/cast.hpp>

#include <vespa/log/log.h>
LOG_SETUP(".document.update.fieldpathupdate");

using vespalib::nbostream;

namespace document {

using namespace fieldvalue;

IMPLEMENT_IDENTIFIABLE(AssignFieldPathUpdate, FieldPathUpdate);

AssignFieldPathUpdate::AssignFieldPathUpdate()
    : FieldPathUpdate(),
      _newValue(),
      _expression(),
      _removeIfZero(false),
      _createMissingPath(false)
{ }


AssignFieldPathUpdate::AssignFieldPathUpdate(
        const DataType& type,
        stringref fieldPath,
        stringref whereClause,
        const FieldValue& newValue)
    : FieldPathUpdate(fieldPath, whereClause),
      _newValue(newValue.clone()),
      _expression(),
      _removeIfZero(false),
      _createMissingPath(true)
{
    checkCompatibility(*_newValue, type);
}

AssignFieldPathUpdate::AssignFieldPathUpdate(stringref fieldPath, stringref whereClause, stringref expression)
    : FieldPathUpdate(fieldPath, whereClause),
      _newValue(),
      _expression(expression),
      _removeIfZero(false),
      _createMissingPath(true)
{
    if (_expression.empty()) {
        throw vespalib::IllegalArgumentException("Cannot create an arithmetic "
                "assignment update with an empty expression", VESPA_STRLOC);
    }
}

AssignFieldPathUpdate::~AssignFieldPathUpdate() { }

FieldPathUpdate*
AssignFieldPathUpdate::clone() const {
    return new AssignFieldPathUpdate(*this);
}
namespace {

class AssignValueIteratorHandler : public IteratorHandler
{
public:
    AssignValueIteratorHandler(const FieldValue& newValue,
                               bool removeIfZero,
                               bool createMissingPath_)
            : _newValue(newValue), _removeIfZero(removeIfZero),
              _createMissingPath(createMissingPath_)
    {}

    ModificationStatus doModify(FieldValue& fv) override;
    bool onComplex(const Content&) override { return false; }
    bool createMissingPath() const override { return _createMissingPath; }

private:
    const FieldValue& _newValue;
    bool _removeIfZero;
    bool _createMissingPath;
};

class AssignExpressionIteratorHandler : public IteratorHandler
{
public:
    AssignExpressionIteratorHandler(
            const DocumentTypeRepo& repo,
            Document& doc,
            const vespalib::string& expression,
            bool removeIfZero,
            bool createMissingPath_)
            : _calc(repo, expression),
              _doc(doc),
              _removeIfZero(removeIfZero),
              _createMissingPath(createMissingPath_)
    {}

    ModificationStatus doModify(FieldValue& fv) override;
    bool onComplex(const Content&) override { return false; }
    bool createMissingPath() const override { return _createMissingPath; }

private:
    DocumentCalculator _calc;
    Document& _doc;
    bool _removeIfZero;
    bool _createMissingPath;
};

ModificationStatus
AssignValueIteratorHandler::doModify(FieldValue& fv) {
    LOG(spam, "fv = %s", fv.toString().c_str());
    if (!(*fv.getDataType() == *_newValue.getDataType())) {
        vespalib::string err = vespalib::make_string(
                "Trying to assign \"%s\" of type %s to an instance of type %s",
                _newValue.toString().c_str(), _newValue.getClass().name(),
                fv.getClass().name());
        throw vespalib::IllegalArgumentException(err, VESPA_STRLOC);
    }
    if (_removeIfZero
        && _newValue.inherits(NumericFieldValueBase::classId)
        && static_cast<const NumericFieldValueBase&>(_newValue).getAsLong() == 0)
    {
        return ModificationStatus::REMOVED;
    }
    fv.assign(_newValue);
    return ModificationStatus::MODIFIED;
}

ModificationStatus
AssignExpressionIteratorHandler::doModify(FieldValue& fv) {
    LOG(spam, "fv = %s", fv.toString().c_str());
    if (fv.inherits(NumericFieldValueBase::classId)) {
        std::unique_ptr<select::VariableMap> varHolder = std::make_unique<select::VariableMap>();
        select::VariableMap & vars = *varHolder;
        for (VariableMap::const_iterator i(getVariables().begin()),
                 e(getVariables().end()); i != e; ++i)
        {
            if (i->second.key.get() && i->second.key->inherits(NumericFieldValueBase::classId)) {
                vars[i->first] = i->second.key->getAsDouble();
            } else {
                vars[i->first] = i->second.index;
            }
        }

        vars["value"] = fv.getAsDouble();

        try {
            double res = _calc.evaluate(_doc, std::move(varHolder));
            if (_removeIfZero && static_cast<uint64_t>(res) == 0) {
                return ModificationStatus::REMOVED;
            } else {
                fv.assign(DoubleFieldValue(res));
            }
        } catch (const vespalib::IllegalArgumentException&) {
            // Divide by zero does not modify the document field
            return ModificationStatus::NOT_MODIFIED;
        } catch (const boost::bad_numeric_cast&) {
            // Underflow/overflow does not modify
            return ModificationStatus::NOT_MODIFIED;
        }
    } else {
        throw vespalib::IllegalArgumentException(
                vespalib::make_string("Trying to perform arithmetic on %s of type %s",
                                      fv.toString().c_str(), fv.getDataType()->toString().c_str()),
                VESPA_STRLOC);
    }
    return ModificationStatus::MODIFIED;
}

}

std::unique_ptr<IteratorHandler>
AssignFieldPathUpdate::getIteratorHandler(Document& doc, const DocumentTypeRepo & repo) const
{
    if (!_expression.empty()) {
        return std::make_unique<AssignExpressionIteratorHandler>(repo, doc, _expression, _removeIfZero, _createMissingPath);
    } else {
        return std::make_unique<AssignValueIteratorHandler>(*_newValue, _removeIfZero, _createMissingPath);
    }
}

bool
AssignFieldPathUpdate::operator==(const FieldPathUpdate& other) const
{
    if (other.getClass().id() != AssignFieldPathUpdate::classId) return false;
    if (!FieldPathUpdate::operator==(other)) return false;
    const AssignFieldPathUpdate& assignOther
        = static_cast<const AssignFieldPathUpdate&>(other);
    if (assignOther._newValue.get() && _newValue.get()) {
        if (*assignOther._newValue != *_newValue) return false;
    }
    // else: should always have at least 1 with non-empty expression
    return (assignOther._expression == _expression)
            && (assignOther._removeIfZero == _removeIfZero)
            && (assignOther._createMissingPath == _createMissingPath);
}

void
AssignFieldPathUpdate::print(std::ostream& out, bool verbose, const std::string& indent) const
{
    out << "AssignFieldPathUpdate(\n";
    FieldPathUpdate::print(out, verbose, indent + "  ");
    if (_newValue.get()) {
        out << ",\n" << indent << "  " << "newValue=";
        _newValue->print(out, verbose, indent + "  ");
    } else {
        out << ",\n" << indent << "  " << "expression='" << _expression << "'";
    }
    out << ", removeIfZero=" << (_removeIfZero ? "yes" : "no")
        << ", createMissingPath=" << (_createMissingPath ? "yes" : "no")
        << "\n" << indent << ")";
}

void
AssignFieldPathUpdate::deserialize(const DocumentTypeRepo& repo, const DataType& type, nbostream & stream)
{
    FieldPathUpdate::deserialize(repo, type, stream);

    uint8_t flags = 0x00;
    stream >> flags;

    _removeIfZero = (flags & REMOVE_IF_ZERO) != 0;
    _createMissingPath = (flags & CREATE_MISSING_PATH) != 0;

    if (flags & ARITHMETIC_EXPRESSION) {
        _expression = getString(stream);
    } else {
        FieldPath path;
        type.buildFieldPath(path, getOriginalFieldPath());
        _newValue.reset(getResultingDataType(path).createFieldValue().release());
        VespaDocumentDeserializer deserializer(repo, stream, Document::getNewestSerializationVersion());
        deserializer.read(*_newValue);
    }
}

} // ns document

File: document/src/main/java/com/yahoo/document/fieldpathupdate/AssignFieldPathUpdate.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.document.fieldpathupdate;

import com.yahoo.document.Document;
import com.yahoo.document.DocumentCalculator;
import com.yahoo.document.DocumentType;
import com.yahoo.document.datatypes.FieldPathIteratorHandler;
import com.yahoo.document.datatypes.FieldValue;
import com.yahoo.document.datatypes.NumericFieldValue;
import com.yahoo.document.select.parser.ParseException;
import com.yahoo.document.serialization.DocumentUpdateReader;
import com.yahoo.document.serialization.VespaDocumentSerializer6;

import java.util.HashMap;
import java.util.Map;

/**
 * @author Thomas Gundersen
 */
public class AssignFieldPathUpdate extends FieldPathUpdate {

    class SimpleAssignIteratorHandler extends FieldPathIteratorHandler {
        FieldValue newValue;
        boolean removeIfZero;
        boolean createMissingPath;

        SimpleAssignIteratorHandler(FieldValue newValue, boolean removeIfZero, boolean createMissingPath) {
            this.newValue = newValue;
            this.removeIfZero = removeIfZero;
            this.createMissingPath = createMissingPath;
        }

        @Override
        public ModificationStatus doModify(FieldValue fv) {
            if (!fv.getDataType().equals(newValue.getDataType())) {
                throw new IllegalArgumentException("Trying to assign " + newValue + " of type " + newValue.getDataType() + " to an instance of " + fv.getDataType());
            } else {
                if (removeIfZero && (newValue instanceof NumericFieldValue) && ((NumericFieldValue)newValue).getNumber().longValue() == 0) {
                    return ModificationStatus.REMOVED;
                }
                fv.assign(newValue);
            }
            return ModificationStatus.MODIFIED;
        }

        @Override
        public boolean createMissingPath() {
            return createMissingPath;
        }

        @Override
        public boolean onComplex(FieldValue fv) {
            return false;
        }
    }

    class MathAssignIteratorHandler extends FieldPathIteratorHandler {
        DocumentCalculator calc;
        Document doc;
        boolean removeIfZero;
        boolean createMissingPath;

        MathAssignIteratorHandler(String expression, Document doc, boolean removeIfZero, boolean createMissingPath) throws ParseException {
            this.calc = new DocumentCalculator(expression);
            this.doc = doc;
            this.removeIfZero = removeIfZero;
            this.createMissingPath = createMissingPath;
        }

        @Override
        public ModificationStatus doModify(FieldValue fv) {
            if (fv instanceof NumericFieldValue) {
                Map<String, Object> vars = new HashMap<String, Object>();
                for (Map.Entry<String, IndexValue> entry : getVariables().entrySet()) {
                    if (entry.getValue().getKey() != null && entry.getValue().getKey() instanceof NumericFieldValue) {
                        vars.put(entry.getKey(), ((NumericFieldValue)entry.getValue().getKey()).getNumber());
                    } else {
                        vars.put(entry.getKey(), entry.getValue().getIndex());
                    }
                }
                vars.put("value", ((NumericFieldValue)fv).getNumber());

                try {
                    Number d = calc.evaluate(doc, vars);
                    if (removeIfZero && d.longValue() == 0) {
                        return ModificationStatus.REMOVED;
                    } else {
                        fv.assign(calc.evaluate(doc, vars));
                    }
                } catch (IllegalArgumentException e) {
                    // Ignore divide by zero
                    return ModificationStatus.NOT_MODIFIED;
                }
            } else {
                throw new IllegalArgumentException("Trying to perform arithmetic on " + fv + " of type " + fv.getDataType());
            }
            return ModificationStatus.MODIFIED;
        }

        @Override
        public boolean createMissingPath() {
            return createMissingPath;
        }

        @Override
        public boolean onComplex(FieldValue fv) {
            return false;
        }
    }

    FieldValue fieldValue = null;
    String expression = null;
    boolean createMissingPath = true;
    boolean removeIfZero = false;

    // Flag bits
    public static final int ARITHMETIC_EXPRESSION = 1;
    public static final int REMOVE_IF_ZERO = 2;
    public static final int CREATE_MISSING_PATH = 4;

    /**
     * Creates an assignment update that overwrites the old value with the given new value.
     *
     * @param type The document type the assignment works on.
     * @param fieldPath The field path of the field to be overwritten.
     * @param whereClause A document selection string that selects documents and variables to be updated.
     * @param newValue The new value of the assignment.
     */
    public AssignFieldPathUpdate(DocumentType type, String fieldPath, String whereClause, FieldValue newValue) {
        super(FieldPathUpdate.Type.ASSIGN, type, fieldPath, whereClause);
        setNewValue(newValue);
    }

    public AssignFieldPathUpdate(DocumentType type, String fieldPath, FieldValue newValue) {
        super(FieldPathUpdate.Type.ASSIGN, type, fieldPath, null);
        setNewValue(newValue);
    }

    /**
     * Creates an assign statement based on a mathematical expression.
     *
     * @param type The document type the assignment works on.
     * @param fieldPath The field path of the field to be overwritten.
     * @param whereClause A document selection string that selects documents and variables to be updated.
     * @param expression The mathematical expression to apply. Use $value to signify the previous value of the field.
     */
    public AssignFieldPathUpdate(DocumentType type, String fieldPath, String whereClause, String expression) {
        super(FieldPathUpdate.Type.ASSIGN, type, fieldPath, whereClause);
        setExpression(expression);
    }

    /**
     * Creates an assign update from a serialized object.
     *
     * @param type The document type the assignment will work on.
     * @param reader A reader that can deserialize something into this object.
     */
    public AssignFieldPathUpdate(DocumentType type, DocumentUpdateReader reader) {
        super(FieldPathUpdate.Type.ASSIGN, type, reader);
        reader.read(this);
    }

    public AssignFieldPathUpdate(DocumentType type, String fieldPath) {
        super(FieldPathUpdate.Type.ASSIGN, type, fieldPath, null);
    }

    /**
     * Turns this assignment into a literal one.
     *
     * @param value The new value to assign to the document.
     */
    public void setNewValue(FieldValue value) {
        fieldValue = value;
        expression = null;
    }

    /**
     *
     * @return Returns the value to assign, or null if this is a mathematical expression.
     */
    public FieldValue getNewValue() {
        return fieldValue;
    }

    /**
     * Turns this assignment into a mathematical expression assignment.
     *
     * @param value The expression to use for assignment.
     */
    public void setExpression(String value) {
        expression = value;
        fieldValue = null;
    }

    /**
     *
     * @return Returns the arithmetic expression to assign, or null if this is not a mathematical expression.
     */
    public String getExpression() {
        return expression;
    }

    /**
     * If set to true, and the new value assigned evaluates to a numeric value of 0, removes the value instead of setting it.
     * Default is false.
     */
    public void setRemoveIfZero(boolean removeIfZero) {
        this.removeIfZero = removeIfZero;
    }

    /**
     * If set to true, and any part of the field path specified does not exist (except for array indexes), we create the path as necessary.
     * Default is true.
     */
    public void setCreateMissingPath(boolean createMissingPath) {
        this.createMissingPath = createMissingPath;
    }

    /**
     *
     * @return Returns true if this assignment is an arithmetic operation.
     */
    public boolean isArithmetic() {
        return expression != null;
    }

    FieldPathIteratorHandler getIteratorHandler(Document doc) {
        if (expression != null) {
            try {
                return new MathAssignIteratorHandler(expression, doc, removeIfZero, createMissingPath);
            } catch (ParseException e) {
                return null;
            }
        } else {
            return new SimpleAssignIteratorHandler(fieldValue, removeIfZero, createMissingPath);
        }
    }

    @Override
    public void serialize(VespaDocumentSerializer6 data) {
        data.write(this);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;

        AssignFieldPathUpdate that = (AssignFieldPathUpdate) o;

        if (createMissingPath != that.createMissingPath) return false;
        if (removeIfZero != that.removeIfZero) return false;
        if (expression != null ? !expression.equals(that.expression) : that.expression != null) return false;
        if (fieldValue != null ? !fieldValue.equals(that.fieldValue) : that.fieldValue != null) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = super.hashCode();
        result = 31 * result + (fieldValue != null ? fieldValue.hashCode() : 0);
        result = 31 * result + (expression != null ? expression.hashCode() : 0);
        result = 31 * result + (createMissingPath ? 1 : 0);
        result = 31 * result + (removeIfZero ? 1 : 0);
        return result;
    }

    @Override
    public String toString() {
        return "Assign: " + super.toString() + " : " + (isArithmetic() ? getExpression() : getNewValue().toString());
    }

    public boolean getCreateMissingPath() {
        return createMissingPath;
    }

    public boolean getRemoveIfZero() {
        return removeIfZero;
    }

    public FieldValue getFieldValue() {
        return fieldValue;
    }

}
